rust_src/src/window_configuration.rs

```diff
@@ -0,0 +1,203 @@
+            && self.pixel_height.eq(other.pixel_height)
+            && self.pixel_width.eq(other.pixel_width)
+            && self.left_col.eq(other.left_col)
+            && self.top_line.eq(other.top_line)
+            && self.total_cols.eq(other.total_cols)
+            && self.total_lines.eq(other.total_lines)
+            && self.display_table.eq(other.display_table))
+        {
+            return false;
+        }
+
+        // The next two check the window structure for equality.
+        if !(self.parent.eq(other.parent) && self.prev.eq(other.prev)) {
+            return false;
+        }
+
+        if !(ignore_positions
+            || (self.hscroll.eq(other.hscroll)
+                && self.min_hscroll.eq(other.min_hscroll)
+                && self.start_at_line_beg.eq(other.start_at_line_beg)
+                && equal(self.start, other.start))
+                && equal(self.pointm, other.pointm))
+        {
+            return false;
+        }
+
+        if !(self.left_margin_cols.eq(other.left_margin_cols)
+            && self.right_margin_cols.eq(other.right_margin_cols)
+            && self.left_fringe_width.eq(other.left_fringe_width)
+            && self.right_fringe_width.eq(other.right_fringe_width)
+            && self
+                .fringes_outside_margins
+                .eq(other.fringes_outside_margins)
+            && self.scroll_bar_width.eq(other.scroll_bar_width)
+            && self.scroll_bar_height.eq(other.scroll_bar_height)
+            && self
+                .vertical_scroll_bar_type
+                .eq(other.vertical_scroll_bar_type)
+            && self
+                .horizontal_scroll_bar_type
+                .eq(other.horizontal_scroll_bar_type))
+        {
+            return false;
+        }
+
+        true
+    }
+}
+
+impl From<LispObject> for SaveWindowDataRef {
+    fn from(o: LispObject) -> Self {
+        o.as_window_configuration_or_error()
+    }
+}
+
+impl LispObject {
+    pub fn as_window_configuration(self) -> Option<SaveWindowDataRef> {
+        self.as_vectorlike()
+            .and_then(|v| v.as_window_configuration())
+    }
+
+    pub fn as_window_configuration_or_error(self) -> SaveWindowDataRef {
+        self.as_window_configuration()
+            .unwrap_or_else(|| wrong_type!(Qwindow_configuration_p, self))
+    }
+}
+
+/// Return t if OBJECT is a window-configuration object.
+#[lisp_fn]
+pub fn window_configuration_p(object: LispObject) -> bool {
+    object.as_window_configuration().is_some()
+}
+
+/// Return the frame that CONFIG, a window-configuration object, is about.
+#[lisp_fn]
+pub fn window_configuration_frame(config: SaveWindowDataRef) -> LispFrameRef {
+    let saved_windows = config.saved_windows.as_vector().unwrap();
+    let obj = saved_windows.get(0);
+    let saved = SavedWindowRef::new(obj.as_vector().unwrap().as_mut() as *mut saved_window);
+    saved.window.as_window_or_error().frame.as_frame_or_error()
```

Comment:
Because we have been returning Rust types where possible. This allows the Rust code to use the functions directly.

Reply:
If we know for sure the return value is only ever used by lisp then the type casting is usually not worth it. As we port more of the code we can use the ported DEFUNs more readily than the C could.

Reply:
I'm looking forward to see result types.

Reply:
It will be an interesting experiment for sure. I would prefer them as well but I am not certain we can go that direction. As @Wilfred commented, the elisp stack trace is part of the API. We can't break it.

Reply:
Ah, I forgot about that. So when we would have ported all C code we still couldn't do it without breaking compatibility ?

Reply:
Quite likely. But once we get there we can experiment. Decide what matters or how we want to implement it.